runtime/buffer.cc

```diff
@@ -266,18 +266,27 @@ void copy_impl(raw_buffer& src, raw_buffer& dst, const void* padding) {
     const index_t pad_before = (src_dim0.begin() - dst_dim0.begin()) * elem_size;
     const index_t pad_after = (dst_dim0.end() - src_dim0.end()) * elem_size;
     const index_t size = padded_size - pad_before - pad_after;
-    dst.slice(0);
-    src.slice(0);
 
     constant_buffer buffer;
     if (padding) {
       optimize_fill_value(padding, elem_size, buffer);
+    } else if (src_dim0.empty()) {
+      // src is empty and there is no padding -> nothing to do
+      return;
     } else {
       assert(size == padded_size);
       assert(pad_before == 0);
       assert(pad_after == 0);
     }
 
+    dst.slice(0);
+    src.slice(0);
+
+    if (!padding && src.rank > 0 && src.dim(0).empty()) {
```

Comment:
This shouldn't be necessary, it should be handled properly by `for_each_element`, we need to fix that if not. This also doesn't really make sense, it's only checking one of N possible dimensions.